app/src/main/java/com/github/geohunt/app/data/repository/ClaimRepository.kt

```diff
@@ -0,0 +1,184 @@
+        claimerId = uid!!,
+        claimDate = DateUtils.localFromUtcIso8601(time!!),
+        distance = distance,
+        awardedPoints = awardedPoints,
+        photoUrl = photoUrl!!
+    )
+
+    /**
+     * Retrieve a list of all claims id for a specific user, useful when lazy loading
+     */
+    override suspend fun getClaimIdByUser(user: User): List<String> {
+        require(user.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            database.getReference("claimsByUser")
+                .child(user.id)
+                .get()
+                .await()
+                .run {
+                    children.mapNotNull { it.getValue(String::class.java) }
+                }
+        }
+    }
+
+    /**
+     * Check whether the currently logged user claim the given challenges
+     */
+    override suspend fun doesClaims(challenge: Challenge) : Boolean = withContext(ioDispatcher) {
+        authRepository.requireLoggedIn()
+        val currentUser = authRepository.getCurrentUser()
+
+        getClaimsByUser(currentUser)
+            .any { it.parentChallengeId == challenge.id }
+    }
+
+    /**
+     * Retrieve the score for a given user
+     */
+    override suspend fun getScoreFromUser(user: User) : Long {
+        require(user.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            getClaimsByUser(user).sumOf { it.awardedPoints }
+        }
+    }
+
+    /**
+     * Get all claims of a specific user [user]. If one of his claim is not within the database
+     * due to some internal issues then throws [ClaimNotFoundException]. Notice that this function
+     * does not check whether the provided user exists or not !!
+     */
+    override suspend fun getClaimsByUser(user: User) : List<Claim> {
+        require(user.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            getClaimIdByUser(user).run {
+                map { claimId ->
+                    database.getReference("claims/$claimId").get().asDeferred()
+                }.awaitAll().zip(this).map {
+                    if (!it.first.exists()) {
+                        throw ClaimNotFoundException(it.second)
+                    }
+                    it.first.getValue(FirebaseClaim::class.java)!!.asExternalModel(it.second)
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieve a list of all claims associated with the current challenges
+     */
+    override suspend fun getClaimsByChallenge(challenge: Challenge): List<Claim> {
+        require(challenge.id.isNotEmpty())
+
+        return withContext(ioDispatcher) {
+            database.getReference("claims/${challenge.id}")
+                .get()
+                .await()
+                .run {
+                    children.mapNotNull {
+                        val value = it.getValue(FirebaseClaim::class.java)
+                        if (value != null) it.key!! to value
+                        else null

```

Comment:
A `when` should be clearer:
```suggestion
                        when(val value = it.getValue(FirebaseClaim::class.java)) {
                            null -> null
                            else -> it.key!! to value
                        }
```